Thread-safe malloc

ISSUE: Several Arduino library uses malloc which is not thread-safe, this could lead to heap corruption.

implement thread safe malloc() using __malloc_lock()  __malloc_unlock() with recursive mutex. Note: may need to upgrade toolchain to gcc 6 2017q1

reference
- https://answers.launchpad.net/gcc-arm-embedded/+question/229509
- https://developer.mbed.org/users/AdamGreen/notebook/newlib-nano/
- https://github.com/spark/firmware/issues/826

using linker --wrap option, we can replace fully replace malloc with freeRTOS compatible one 

```
--wrap symbol
Use a wrapper function for symbol. Any undefined reference to symbol will be resolved to __wrap_symbol. Any undefined reference to __real_symbol will be resolved to symbol. This can be used to provide a wrapper for a system function. The wrapper function should be called __wrap_symbol. If it wishes to call the system function, it should call __real_symbol. Here is a trivial example:
void *
__wrap_malloc (int c)
{
  printf ("malloc called with %ld\n", c);
  return __real_malloc (c);
}
If you link other code with this file using --wrap malloc, then all calls to malloc will call the function __wrap_malloc instead. The call to __real_malloc in __wrap_malloc will call the real malloc function. You may wish to provide a __real_malloc function as well, so that links without the --wrap option will succeed. If you do this, you should not put the definition of __real_malloc in the same file as __wrap_malloc; if you do, the assembler may resolve the call before the linker has a chance to wrap it to malloc.
```

--- a/cores/nRF5/freertos/Source/portable/MemMang/heap_3.c
+++ b/cores/nRF5/freertos/Source/portable/MemMang/heap_3.c
@@ -55,13 +55,28 @@ task.h is included from an application file. */
 
 /*-----------------------------------------------------------*/
 
+// link to libnano's malloc
+// require "-Wl,--wrap=malloc -Wl,--wrap=free" linker option
+extern void *__real_malloc(size_t size);
+extern void __real_free(void *ptr);
+
+void* __wrap_malloc (size_t c)
+{
+  return (xTaskGetSchedulerState() == taskSCHEDULER_RUNNING) ? pvPortMalloc(c) : __real_malloc(c);
+}
+
+void __wrap_free (void *ptr)
+{
+  return (xTaskGetSchedulerState() == taskSCHEDULER_RUNNING) ? vPortFree(ptr) : __real_free(ptr);
+}
+
 void *pvPortMalloc( size_t xWantedSize )
 {
 void *pvReturn;
 
 	vTaskSuspendAll();
 	{
-		pvReturn = malloc( xWantedSize );
+		pvReturn = __real_malloc( xWantedSize );
 		traceMALLOC( pvReturn, xWantedSize );
 	}
 	( void ) xTaskResumeAll();
@@ -86,7 +101,7 @@ void vPortFree( void *pv )
 	{
 		vTaskSuspendAll();
 		{
-			free( pv );
+			__real_free( pv );
 			traceFREE( pv, 0 );
 		}
 		( void ) xTaskResumeAll();

--- a/cores/nRF5/freertos/config/FreeRTOSConfig.h
+++ b/cores/nRF5/freertos/config/FreeRTOSConfig.h
@@ -184,15 +184,15 @@ standard names - or at least those used in the unmodified vector table. */
 
 /** Implementation note:  Use this with caution and set this to 1 ONLY for debugging
  * ----------------------------------------------------------
-     * Set the value of configUSE_DISABLE_TICK_AUTO_CORRECTION_DEBUG to below for enabling or disabling RTOS tick auto correction:
-     * 0. This is default. If the RTC tick interrupt is masked for more than 1 tick by higher priority interrupts, then most likely
-     *    one or more RTC ticks are lost. The tick interrupt inside RTOS will detect this and make a correction needed. This is needed
-     *    for the RTOS internal timers to be more accurate.
-     * 1. The auto correction for RTOS tick is disabled even though few RTC tick interrupts were lost. This feature is desirable when debugging
-     *    the RTOS application and stepping though the code. After stepping when the application is continued in debug mode, the auto-corrections of
-     *    RTOS tick might cause asserts. Setting configUSE_DISABLE_TICK_AUTO_CORRECTION_DEBUG to 1 will make RTC and RTOS go out of sync but could be
-     *    convenient for debugging.
-     */
+ * Set the value of configUSE_DISABLE_TICK_AUTO_CORRECTION_DEBUG to below for enabling or disabling RTOS tick auto correction:
+ * 0. This is default. If the RTC tick interrupt is masked for more than 1 tick by higher priority interrupts, then most likely
+ *    one or more RTC ticks are lost. The tick interrupt inside RTOS will detect this and make a correction needed. This is needed
+ *    for the RTOS internal timers to be more accurate.
+ * 1. The auto correction for RTOS tick is disabled even though few RTC tick interrupts were lost. This feature is desirable when debugging
+ *    the RTOS application and stepping though the code. After stepping when the application is continued in debug mode, the auto-corrections of
+ *    RTOS tick might cause asserts. Setting configUSE_DISABLE_TICK_AUTO_CORRECTION_DEBUG to 1 will make RTC and RTOS go out of sync but could be
+ *    convenient for debugging.
+ */
 #define configUSE_DISABLE_TICK_AUTO_CORRECTION_DEBUG     0
 
 // Sysview is enabled by default at debug level 3

--- a/cores/nRF5/new.cpp
+++ b/cores/nRF5/new.cpp
@@ -76,16 +76,4 @@ caddr_t _sbrk( int incr )
   return (caddr_t) prev_heap;
 }
 
-void __malloc_lock(struct _reent *ptr)
-{
-  (void) ptr;
-  vTaskSuspendAll();
-}
-
-void __malloc_unlock(struct _reent *ptr)
-{
-  (void) ptr;
-  xTaskResumeAll();
-}
-
 }

--- a/cores/nRF5/rtos.h
+++ b/cores/nRF5/rtos.h
@@ -50,8 +50,6 @@
 #include "queue.h"
 #include "semphr.h"
 
-#define DEBUG_MALLOC    1
-
 #define DELAY_FOREVER   portMAX_DELAY
 
 enum
@@ -63,19 +61,13 @@ enum
   TASK_PRIO_HIGHEST = 4,
 };
 
-#define ms2tick              pdMS_TO_TICKS
+#define ms2tick       pdMS_TO_TICKS
+#define tick2ms(tck)  ( ( ((uint64_t)(tck)) * 1000) / configTICK_RATE_HZ )
+#define tick2us(tck)  ( ( ((uint64_t)(tck)) * 1000000) / configTICK_RATE_HZ )
 
-#define tick2ms(tck)         ( ( ((uint64_t)(tck)) * 1000) / configTICK_RATE_HZ )
-#define tick2us(tck)         ( ( ((uint64_t)(tck)) * 1000000) / configTICK_RATE_HZ )
-
-#if DEBUG_MALLOC
-  #define rtos_malloc_type(_type)   ({ LOG_LV2("MALLOC", #_type " = %d bytes", sizeof(_type)); ((_type*) rtos_malloc(sizeof(_type))); })
-#else
-  #define rtos_malloc_type(_type)   ((_type*) rtos_malloc(sizeof(_type)))
-#endif
-
-#define rtos_malloc malloc
-#define rtos_free   free
+// legacy thread-safe malloc/free
+#define rtos_malloc   malloc
+#define rtos_free     free
 
 // Visible only with C++
 #ifdef __cplusplus

--- a/cores/nRF5/rtos.h
+++ b/cores/nRF5/rtos.h
@@ -74,15 +74,8 @@ enum
   #define rtos_malloc_type(_type)   ((_type*) rtos_malloc(sizeof(_type)))
 #endif
 
-static inline void* rtos_malloc(size_t _size)
-{
-  return (xTaskGetSchedulerState() == taskSCHEDULER_NOT_STARTED) ? malloc(_size) : pvPortMalloc(_size);
-}
-
-static inline void rtos_free( void *pv )
-{
-  return (xTaskGetSchedulerState() == taskSCHEDULER_NOT_STARTED) ? free(pv) : vPortFree(pv);
-}
+#define rtos_malloc malloc
+#define rtos_free   free
 
 // Visible only with C++
 #ifdef __cplusplus

--- a/keywords.txt
+++ b/keywords.txt
@@ -36,10 +36,6 @@ delay_ns	KEYWORD2
 analogOversampling	KEYWORD2
 analogReadVDD	KEYWORD2
 
-# RTOS
-rtos_malloc	KEYWORD2
-rtos_free	KEYWORD2
-
 nrf52_flash_init	KEYWORD2
 nrf52_flash_erase_sector	KEYWORD2
 nrf52_flash_write	KEYWORD2

--- a/platform.txt
+++ b/platform.txt
@@ -49,7 +49,7 @@ compiler.elf2bin.flags=-O binary
 compiler.elf2bin.cmd=arm-none-eabi-objcopy
 compiler.elf2hex.flags=-O ihex
 compiler.elf2hex.cmd=arm-none-eabi-objcopy
-compiler.ldflags=-mcpu={build.mcu} -mthumb {build.float_flags} -Wl,--cref -Wl,--check-sections -Wl,--gc-sections -Wl,--unresolved-symbols=report-all -Wl,--warn-common -Wl,--warn-section-align --specs=nano.specs --specs=nosys.specs
+compiler.ldflags=-mcpu={build.mcu} -mthumb {build.float_flags} -Wl,--cref -Wl,--check-sections -Wl,--gc-sections -Wl,--unresolved-symbols=report-all -Wl,--warn-common -Wl,--warn-section-align -Wl,--wrap=malloc -Wl,--wrap=free --specs=nano.specs --specs=nosys.specs
 compiler.size.cmd=arm-none-eabi-size
 
 # this can be overriden in boards.txt

